Updated to use Jenkins API, handles deletion using thresholds.

franostanisicserato opened the pull request:
Consolidated other pull requests, cleaned up some of the code and added a new method for deletion whereby the user specifies an upper bound to trigger the check (as prior), but instead of how it was prior (deleting all workspaces) the plugin now deletes workspaces sequentially, starting with the oldest, until a lower threshold is hit (also specified by the user). The user can also set this to 0 to delete all workspaces as the original functionality did.

Commit:
First commit, consolidating all changes made offline. Now clears using jenkins API and thersholds.

franostanisicserato committed:
First commit, consolidating all changes made offline. Now clears using jenkins API and thresholds.

franostanisicserato committed:
Merge branch 'master' of https://github.com/serato/diskcheck-plugin

franostanisicserato committed:
Forgot to add updated pom. Wwops.

franostanisicserato committed:
Provided functionality for clear all.

franostanisicserato committed:
Code cleanup: indenting, log messages and other cleanups from other pull requests.

batmat commented on pom.xml:
```diff
@@ -31,6 +31,11 @@
+      <id>fsta</id>
+      <name>Frano Stanisic</name>
+      <email>frano.stanisic@serato.com</email>
+    </developer>
```
This block is used for maintainers, IIUC you're not a maintainer on this plugin. If you want your name be added, use \<contributors> instead.

batmat commented on pom.xml:
```diff
@@ -3,12 +3,12 @@
   <groupId>org.jenkins-ci.plugins</groupId>
   <artifactId>diskcheck</artifactId>
-  <version>0.27</version>
+  <version>0.28</version>
```
You shouldn't touch that field in a PR generally. 
Btw, it's not normal it doesn't end with -SNAPSHOT.

-SNAPSHOT should only be absent in tags.

batmat commented on pom.xml:
```diff
@@ -46,6 +51,8 @@
 
+    
+    
   <pluginRepositories>
```
generally don't stage unrelated modification (and especially formatting/whitespaces things).

Why: It's making future history lookup more difficult since your modification is gonna appear as last commit for those lines, when they actually weren't "actual" modifications.

batmat commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
- * 
+ *
  * @author Manoj Thakkar
+ * @author Frano Stanisic (8th Decemeber 2015, updated deletion to use Jenkins API/only delete files up to a threshhold)
```
In general, I'd advise to _not_ add those tags. IMO they're not useful since they become out-of-date very quickly. And SCM is a better tool to know what changed in a source file.

batmat commented:
As you requested review: thanks for the work!

I'm not the maintainer of that plugin, but here's my 2 cents:
- IIUC you merged many PR in that one. I don't think that's gonna work. If you want to update a PR, then comment on it, or submit a new updated one.
- IMO you should aim to have one "goal" per PR
- Use rebase -i to remove commits like "woops", and the likes to get a clear commit log: goal of commits is to make your intent as clear as possible.

Hope this helps

oleg-nenashev commented on pom.xml:
```diff
@@ -3,12 +3,12 @@
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.532</version>
+    <version>1.542</version>
```
Better to stick to Jenkins LTS

oleg-nenashev commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+     *
+     * @param build
+     * @param launcher
+     * @param listener
```
Please don't create Javadoc entries without descriptions. When you try to release the plugin on JDK8, it blows up

oleg-nenashev commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+        }
 
+        Node node1 = build.getBuiltOn();
+        Computer Comp = node1.toComputer();
```
node1 may be null in corner cases

oleg-nenashev commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+        // am not 100% on what it does I have just left it here.
+        if (DiskSpaceMonitor.DESCRIPTOR.get(Comp) == null) {
+            log.println("No slave data available, waiting and retrying ...");
+            Thread.sleep(10000);
```
Please do not use synchronous delays in the main build precess. Makes sense to Asynchronous implementation with Future<>

oleg-nenashev commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+        }
+        // End unsure bit.
 
+        long size = build.getWorkspace().getParent().getUsableDiskSpace();
```
Workspace may be null (corner-case)

oleg-nenashev commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+                    build.getWorkspace().getParent().deleteRecursive();
+                    log.println("All workspaces deleted successfully.");
+                } else {
+                    if (clearUntil < SpaceThreshold) {
```
Maybe makes sense to also have a field checkers in the descriptor

oleg-nenashev commented:
The code is barely reviewable due to the formatting changes. Since the original code is not excellent, it's hard to discover issues in the new code

amuniz commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+ * @author Frano Stanisic (8th Decemeber 2015, updated deletion to use Jenkins API/only delete files up to a threshhold)
  * 
  */
+public class Diskcheck extends BuildWrapper {
```
Using `SimpleBuildWrapper` instead of `BuildWrapper` (by following [this instructions](http://javadoc.jenkins-ci.org/jenkins/tasks/SimpleBuildWrapper.html)) would make this plugin plugin more actual (by using newer Jenkins APIs) and compatible with Workflow (which is make usable as a Workflow step).

amuniz replied:
Note that this is only a suggestion, not required at all, of course.

amuniz commented on src/main/java/org/jenkinsci/plugin/Diskcheck.java:
```diff
@@ -10,195 +11,180 @@
+            if (DiskSpaceMonitor.DESCRIPTOR.get(Comp) == null) {
+                log.println("Could not retrieve slave data, exiting disk check for this slave.");
+            }
+            System.exit(0);
```
Would not this stop the entire JVM (including Jenkins)?

Commit:
Fixed the plugin shutting down the entire Jenkins!!! 

This was happening when a slave was connecting and did not have any disk information, then the plugin was doing a system.exit()...

Commit:
Removed docstring warnings

Removed unused and empty parameters.

Commit:
Fixed potential null pointers

Followed bug/warning from the build to remove the warnings.

Commit:
Removed Jelly unit test errors.

Jelly now needs this line for security reasons.

Commit:
Ignore the target folder from git

As it's the output build.

Commit:
Updated the jenkins plugin version to be able to build

It was failing with weird errors otherwise, also it was quite outdated.

Commit:
Bumped the version to 0.99

We don't want to have the same version as the original plugin to avoid updating our one with the official one (as shown in Jenkins plugin manager) so we make sure it's higher. We will eventually use our own version/id to diverge from the original one.
